Doc/library/__main__.rst

```diff
@@ -1,25 +1,183 @@
+prompt, from standard input, from a file argument, from a :option:`-c`
+argument, or from a :option:`-m` argument, but **not** when it is initialized
+from an import statement.  In any of these situations, the module's
+``__name__`` is set equal to ``'__main__'``.
+
+The only other context in which Python code is run is when it is imported
+through an import statement. In that case, ``__name__`` is set equal to the
+module's name: usually the name of the file without the ``.py`` extension.
+
+As a result, a module can discover whether or not it is running in the
+top-level environment by checking its own ``__name__``, which allows a common
+idiom for conditionally executing code when the module is not initialized from
+an import statement::
+
+    if __name__ == '__main__':
+        # Execute when the module is not initialized from an import statement.
+        ...
+
+
+Idiomatic Usage
+^^^^^^^^^^^^^^^
+
+Putting as few statements as possible in the block below ``if __name___ ==
+'__main__'`` can improve code clarity. Most often, a function named *main*
+encapsulates the program's primary behavior::
+
+    # echo.py
+
+    import shlex
+    import sys
+
+
+    def echo(phrase: str):
+       """A dummy wrapper around print."""
+       # for demonstration purposes, you can imagine that there is some
+       # valuable and reusable logic inside this function
+       print(phrase)
+
+
+    def main():
+        """Echo the input arguments to standard output"""
+        echo(shlex.join(sys.argv))
+
+
+    if __name__ == '__main__':
+        sys.exit(main())  # next section explains the use of sys.exit
+
+This has the added benefit of the *echo* function itself being isolated and
+importable elsewhere. When ``echo.py`` is imported, the ``echo`` and ``main``
+functions will be defined, but neither of them will be called, because
+``__name__ != '__main__'``.
+
+
+Packaging Considerations
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+For detailed documentation on Python packaging, see the
+`Python Packaging User Guide. <https://packaging.python.org/>`_
+
+*main* functions are often used to create command line tools by specifying them
+as entry points for console scripts.  When this is done, pip inserts the
+function call into a template script, where the return value of *main* is
+passed into :func:`sys.exit`. For example::
+
+    sys.exit(main())
+
+Since the call to *main* is wrapped in :func:`sys.exit`, the expectation is
+that your function will return some value acceptable as an input to
+:func:`sys.exit`; typically, an integer or ``None`` (which is implicitly
+returned if your function does not have a return statement).
+
+By proactively following this convention ourselves, our module will have the
+same behavior when run directly (``python3 echo.py``) as it will have if we
+later package it as a console script entry-point in a pip-installable package.
+That is why the ``echo.py`` example from earlier used the ``sys.exit(main())``
```

Comment:
Is it worth reminding readers here that under these circumstances returning a string value will print the message and return an exit code of 1?